runtimes/rust/src/api.rs

```diff
@@ -164,6 +166,32 @@ impl ServiceApiScope {
         self
     }
 
+    pub fn endpoint_new<Q, I, F, R>(&mut self, name: &'static str, handler: F) -> &mut Self
+    where
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(&ServiceApiState<'_>, Q) -> R + 'static + Clone + Send + Sync,
+        R: IntoFuture<Item = I, Error = crate::api::ApiError> + 'static,
+    {
+        let blockchain = self.blockchain.clone();
+        let descriptor = self.descriptor.clone();
+        self.inner
+            .endpoint(name, move |query: Q| -> crate::api::ApiFutureResult<I> {
+                let desc = descriptor.clone();
+                let state = ServiceApiState::from_api_context(
+                    &blockchain,
+                    (desc.0, desc.1.as_ref()).into(),
+                    name,
+                );
+                let result = handler(&state, query);
+                let future = result
+                    .into_future()
+                    .map_err(move |err| err.source(format!("{}:{}", desc.1, desc.0)));
```

Comment:
Does this mean that the error source will be overridden even if the user sets it? If this is intended, I'd suggest to make `ApiError::source()` hidden from docs to not trip the user; if not, then the user-provided source should not be overridden.

Also, it could make sense exploring more structured source presentation, e.g.

```rust
pub struct ApiError {
    // ...
    source: serde_json::Value,
}
```

or even

```rust
pub struct ApiError<S> {
    // ...
    source: S,
}
```

...although this can be done after this PR is merged.